File: Lab3KarpaBohdan36/src/Lab3KarpaBohdan36/Tree.java
package Lab3KarpaBohdan36;

/**
 * Інтерфейс Growable містить метод для зростання.
 */
interface Growable {
    void grow(double increment); // Метод для зростання рослини
}

/**
 * Інтерфейс FruitBearing розширює Growable, додаючи нові методи.
 */
interface FruitBearing extends Growable {
    void bearFruit(); // Метод для плодоношення
    int getFruitCount(); // Метод для отримання кількості плодів
}

/**
 * Клас Tree реалізує інтерфейси Growable і FruitBearing.
 */
public class Tree extends Plant implements FruitBearing {
    private String treeType;
    private int fruitCount;

    public Tree(Root root, Stem stem, Leaves leaves, String treeType) {
        super(root, stem, leaves);
        this.treeType = treeType;
        this.fruitCount = 0; // Кількість плодів спочатку 0
    }

    public String getTreeType() {
        return treeType;
    }

    public void setTreeType(String treeType) {
        this.treeType = treeType;
    }

    @Override
    public String getDescription() {
        return "Дерево типу: " + treeType + "\n" +
                getRoot().toString() + "\n" +
                getStem().toString() + "\n" +
                getLeaves().toString() + "\n" +
                "Кількість плодів: " + fruitCount; // Додано опис для кількості плодів
    }

    @Override
    public void grow(double increment) {
        double newHeight = getStem().getHeight() + increment;
        getStem().setHeight(newHeight);
        System.out.println("Дерево виросло на " + increment + " см, нова висота: " + newHeight + " см.");
        bearFruit(); // Викликаємо метод bearFruit() при зростанні дерева
    }

    @Override
    public void bearFruit() {
        fruitCount += 10; // Наприклад, після кожного зростання з'являється 10 нових плодів
        System.out.println("Дерево принесло 10 нових плодів. Загальна кількість плодів: " + fruitCount);
    }

    @Override
    public int getFruitCount() {
        return fruitCount;
    }
}

File: plants/tree.py
from .plant import Plant


class Tree(Plant):
    """Клас, що представляє дерево, наслідуючи рослину."""

    def __init__(self, root, stem, leaves, tree_type):
        """Ініціалізує дерево з заданим типом."""
        super().__init__(root, stem, leaves)
        self.tree_type = tree_type
        self.fruit_count = 0

    def get_tree_type(self):
        """Повертає тип дерева."""
        return self.tree_type

    def set_tree_type(self, tree_type):
        """Задає новий тип дерева."""
        self.tree_type = tree_type

    def grow(self, increment):
        """Збільшує висоту дерева та додає плоди."""
        new_height = self.get_stem().get_height() + increment
        self.get_stem().set_height(new_height)
        self.bear_fruit()

    def bear_fruit(self):
        """Збільшує кількість плодів на дереві."""
        self.fruit_count += 10

    def get_fruit_count(self):
        """Повертає кількість плодів на дереві."""
        return self.fruit_count

    def get_description(self):
        """Повертає опис дерева."""
        return (f"Дерево типу: {self.tree_type}\n"
                f"{self.root}\n{self.stem}\n{self.leaves}\n"
                f"Кількість плодів: {self.fruit_count}")
